man/ddns-updater.5.pod
#   Copyright
#
#	Copyright (C) 2019 Jari Aalto
#
#   License
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program. If not, see <http://www.gnu.org/licenses/>.
#
#   Description
#
#	To learn what TOP LEVEL sections to use in manual pages,
#	see POSIX/Susv standard and "Utility Description Defaults" at
#	http://www.opengroup.org/onlinepubs/009695399/utilities/xcu_chap01.html#tag_01_11
#
#	This manual page in Perl POD format. Read more at
#	http://perldoc.perl.org/perlpod.html or run command:
#
#	    perldoc perlpod
#
#	To check the syntax:
#
#	    podchecker *.pod
#
#	To create manual:
#
#	    pod2man FILE.N.pod > FILE.N

=pod

=head1 NAME

ddnd-updater - a generic dynamic DNS (DDNS) update client

=head1 SYNOPSIS

  ddns-updater [options]

=head1 DESCRIPTION

Program updates IP address to any service that can be updated using
URL calls, provided that a configuration file exists. Templates
are provided with the programs. Examples:

  http://duckdns.org
  http://dns.he.net
  http://ddnss.de
  ....
  (Add your own. See section CONFIGURATION FILES)

To use a new DDNS service provider, consult their URL API
documentation and write respective configurations file. It will be
used next time the program is called.

=head1 OPTIONS

=over 4

=item B<-c, --config NAME>

Without path, the NAME refets to files in configuration directory.
Without path and without the C<.conf> extension, this translates to NAME.conf
in configuration directory. See FILES section.

This option can be supplied multiple times to select configuration
files to run.

=item B<-f, --force>

Force update even if IP is same.

=item B<-l, --list>

List status of configuration files and exit.

=item B<-L --log>

Display log and exit.

=item B<-s, --status>

Show status and exit.

=item B<-S, --syslog>

Send status to syslog. Can only be used by root and only useful from
cron.

=item B<-t, --test, --dry-run>

Run in test mode. No network update.

=item B<-v, --verbose>

Display verbose messages.

=item B<-V, --Verson>

Display version information and exit.

=item B<-h, --help>

Display short help.

=back

=head1 EXAMPLES

Update and show progress:

    ddns-updater --verbose

Show status, is update needed?

    ddns-updater --status

Run only one configuration file NAME.conf

    ddns-update --verbose --conf NAME[.conf]

=head1 BUGS

The program does not use non-POSIX featues like getopt(1) to parse
options. Due to lack of proper programing language constructs in POSIX
shell, combining short options in form of B<-vf> is not supported. Use
separate options like in B<-f> B<-i>.

Likewise the equal sign notation is not supported, like in B<-conf=NAME>.

=head1 ENVIRONMENT

POSIX

=head1 FILES

=head2 CONFIGURATION FILES

The configuration files are searched from followgin directories.
Only the first one found is used:

  $HOME/.config/ddns-updater
  /etc/ddns-updater

The configuration file requirements are:

  - Must use *.conf extension.
  - Written in POSIX shell.
  - Must define following variables:

    # The complete Update URL.
    # In the $URL, it's possible to use literal WHATSMYIP for the current IP.
    URL=...

    # egrep(1) compatible regexps to check response of the URL call.
    REGEXP_OK=...
    REGEXP_NOCHANGE=...

B<Tip>: name the configuration files
according to domain names like "me.duckdns.org.conf"

=head2 CONFIGURATION FILE FOR THE PRORAM

Not required. Possible to adust e.g. C<CULR_OPTIONS> for timeout,
<URL_WHATSMYIP> to query IP etc.

  /etc/defaults/ddns-updater.conf
  $HOME/.ddns-updater

=head2 GENERATED FILES

Written in the CONFIGURATION DIRECTORY.

  00.ip            Last update - ip address
  00.log           Last update - error log
  00.updated       Last update - YYYY-MM-DD HH:MM"

=head1 STANDARDS

None.

=head1 AVAILABILITY

https://github.com/jaalto/project--ddns-updater

=head1 SEE ALSO

ddclient(1)

=head1 AUTHORS

Program was written by <jari.aalto@cante.net>

Released under license GNU GPL version 2 or (at your option) any
later version. For more information about license, visit
<http://www.gnu.org/copyleft/gpl.html>.

=cut


ddns-updater.sh
# /bin/sh
#
#   Copyright
#
#       Copyright (C) 2019 Jari Aalto <jari.aalto@cante.net>
#
#   License
#
#       This program is free software; you can redistribute it and/or modify
#       it under the terms of the GNU General Public License as published by
#       the Free Software Foundation; either version 2 of the License, or
#       (at your option) any later version.
#
#       This program is distributed in the hope that it will be useful,
#       but WITHOUT ANY WARRANTY; without even the implied warranty of
#       MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
#       GNU General Public License for more details.
#
#       You should have received a copy of the GNU General Public License
#       along with this program. If not, see <http://www.gnu.org/licenses/>.
#
#   Description
#
#       Dynamic DNS (DDNS) update client implemented in POSIX shell script.
#
#       See --help. Configuration files must exist before use.

AUTHOR="Jari Aalto <jari.aalto@cante.net>"
VERSION="2019.0708.1118"
LICENSE="GPL-2+"

HELP="\
Synopsis: $0 [option]

DESCRIPTION
  Updates IP address to service duckdns.org

OPTIONS
  -f, --force    Force update even if IP is same.
  -s, --status   Show status and exit.
  -t, --test     Run in test mode. No real update.
  -v, --verbose  Display verbose output.
  -V, --version  Display version information and exit.
  -h, --help     Display short help.

  Please note that stacking of short options is not supported. E.g.
  -v -f cannot be combined into -vf.

FILES
  User configuration files:

  # For duckdns.org
  $CONF/duckdns.domains  Comma separated list of the subnames (not FQDN)
  $CONF/duckdns.token    The token of account (see your profile)

  # For dns.he.org
  $CONF/henet.domains    Comma separated list of the subnames
  $CONF/henet.pass       The account password

  Internal house keeping files:

  00.ip                  Current ip
  00.updated             contains YYYY-MM-DD HH:MM of last update"

# See https://specifications.freedesktop.org/basedir-spec/basedir-spec-latest.html
if [ "$XDG_CONFIG_HOME" ]; then
    CONFHOME=$XDG_CONFIG_HOME/ddns-updater
else
    CONFHOME=$HOME/.config/ddns-updater
fi

CONF=

for dir in $CONFHOME /etc/ddns-updater
do
    if [ -d "$dir" ]; then
        CONF=$dir
    fi
done

DUCKDNS_FILE_DOMAINS=$CONF/duckdns.domains
DUCKDNS_FILE_TOKEN=$CONF/duckdns.token
DUCKDNS_FILE_LOG=$CONF/duckdns.log
DUCKDNS_URI_VERBOSE="&verbose=true"

HENET_FILE_DOMAINS=$CONF/henet.domains
HENET_FILE_PASS=$CONF/henet.pass
HENET_FILE_LOG=$CONF/henet.log

# Use prefix 00.* to make data files to appear first in ls(1)
FILE_IP=$CONF/00.ip
FILE_TIMESTAMP=$CONF/00.updated

Version ()
{
    echo "$VERSION $LICENSE $AUTHOR $HOMEPAGE"
}

Echo ()
{
    [ "$VERBOSE" ] && echo "$*"
}

Warn ()
{
    echo "$*" >&2
}

Die ()
{
    Warn "$*"
    exit 1
}

Date ()
{
    date "+%Y-%m-%d %H:%M"
}

OldIP ()
{
    cat $FILE_IP 2> /dev/null
}

CurrentIP ()
{
    curl --silent ifconfig.co
}

IsHenet ()
{
    [ -f "$HENET_FILE_PASS" ]
}

HenetStatus ()
{
    [ -f "$HENET_FILE_LOG" ] || return 2

    if [ "$VERBOSE" ]; then
        cat $HENET_FILE_LOG
    fi

    grep "^OK" $HENET_FILE_LOG > /dev/null 2>&1
}

Henet ()
{
    :
}

IsDuckdns ()
{
    [ -f "$DUCKDNS_FILE_TOKEN" ]
}

DuckdnsStatus ()
{
    [ -f "$DUCKDNS_FILE_LOG" ] || return 2

    grep "^OK" $DUCKDNS_FILE_LOG > /dev/null 2>&1
}

Duckdns ()
{
    domains=$(sed -e 's/[ \t]*//' $DUCKDNS_FILE_DOMAINS)
    token=$(cat $DUCKDNS_FILE_TOKEN)

    if [ ! "$token" ] ; then
        Die "ERROR: No token id in: $DUCKDNS_FILE_TOKEN"
    fi

    if [ ! "$domains" ]; then
        Die "ERROR: No subdomains in: $DUCKDNS_FILE_DOMAINS"
    fi

    if grep "\." $DUCKDNS_FILE_DOMAINS ; then
        Die "ERROR: FQDN names not allowed, only subdomains" \
            "names in: $DUCKDNS_FILE_DOMAINS"
    fi

    url="https://www.duckdns.org/update?domains=$domains&token=$token&ip=$ip$DUCKDNS_URI_VERBOSE"

    Echo "Info: Updating Duckdns..."
    ${TEST:+echo} curl --silent --insecure --output $DUCKDNS_LOG "$url"

    # Add missing last NEWLINE
    [ "$TEST" ] || echo >> $DUCKDNS_LOG

    if [ "$VERBOSE" ]; then
        # Delete empty lines
        sed '/^[ \t]*$/d' $DUCKDNS_FILE_LOG
    fi

    Echo "Info: Updating Duckdns...done"
}

Main ()
{
    unset TEST

    while :
    do
        case "$1" in
            -f | --force)
                shift
                FORCE=force
                ;;
            -s | --status)
                shift
                status=status
                ;;
            -t | --test | --dry-run)
                shift
                echo "** Running in test mode, no network calls"
                VERBOSE=verbose
                TEST=test
                ;;
            -v | --verbose)
                shift
                VERBOSE=verbose
                ;;
            -V | --version)
                shift
                Version
                return 0
                ;;
            -h | --help)
                shift
                echo "$HELP"
                return 0
                ;;
            -*) Warn "WARN: Unknown option: $1"
                shift
                ;;
            --) shift
                break
                ;;
            *)  break
                ;;
        esac
    done

    if [ ! "$CONF" ] ; then
        Die "ERROR: No configuration directory: $CONFHOME"
    fi

    if [ ! -d "$CONF" ]; then
        Die "ERROR: No configuration directory: $CONF"
    fi

    ip_old=$(OldIP)
    ip=$(CurrentIP)

    Echo "IP old: $ip_old"
    Echo "IP now: $ip"

    if [ "$status" ]; then
        date=$(cat $FILE_TIMESTAMP 2> /dev/null)
        str=" Upated: $date"

        if [ ! "$date" ]; then
            str=" Updated: UNKNOWN (timestamp not available until next update)"
        fi

        if [ "$ip_old" = "$ip" ]; then
            echo "OK IP: $ip (update not needed)$str"
        else
            if ["$ip_old" ]; then
                ip_old="was $ip_old"
            else
                ip_old="old ip UNKNOWN"
            fi

            echo "NOK IP: '$ip' (update needed, $old_ip).$str"
        fi

        return 0
    fi

    if [ ! "$FORCE" ] && [ "$ip_old" = "$ip" ]; then
        Echo "Info: Nothing to update"
        return 0
    else
        echo $ip > $FILE_IP
        Date > $FILE_TIMESTAMP

        done=
        status=0

        IsDuckdns && { Duckdns $ip ; DuckdnsStatus; status=$? ; done=done ;}
        IsHenet   && { Henet $ip ; HenetStatus; status=$? ; done=done ; }

        if [ ! "$done" ]; then
            Die "WARN: No DDNS configuration files. See --help."
        fi

        return $status
    fi
}

Main "$@"

# End of file
